src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -402,41 +380,400 @@ def simpForall (e : Expr) : SimpM Result := withParent e do
+          if bodyTypeFVarIds.contains fvars[idx].fvarId! then
+            deps.insert idx
+          else
+            deps
+        return { info with bodyDeps, bodyTypeDeps, body, bodyType, level }
+
+/--
+Computes which `have`s in the telescope are fixed and which are unused.
+The length of the unused array may be less than the number of `have`s: use `unused.getD i true`.
+-/
+def HaveTelescopeInfo.computeFixedUsed (info : HaveTelescopeInfo) (keepUnused : Bool) :
+    MetaM (Std.HashSet Nat × Array Bool) := do
+  if keepUnused then
+    return (info.bodyTypeDeps, #[])
+  let numHaves := info.haveInfo.size
+  let updateArrayFromBackDeps (arr : Array Bool) (s : Std.HashSet Nat) : Array Bool :=
+    s.fold (init := arr) fun arr idx => arr.set! idx true
+  let mut used : Array Bool := Array.replicate numHaves false
+  -- Initialize `used` with the body's dependencies.
+  -- There is no need to consider `info.bodyTypeDeps` in this computation.
+  used := updateArrayFromBackDeps used info.bodyDeps
+  -- For each used `have`, in reverse order, update `used`.
+  for i in [0:numHaves] do
+    let idx := numHaves - i - 1
+    if used[idx]! then
+      let hinfo := info.haveInfo[idx]!
+      used := updateArrayFromBackDeps used hinfo.typeBackDeps
+      used := updateArrayFromBackDeps used hinfo.valueBackDeps
+  return (info.bodyTypeDeps, used)
+
+/--
+Auxiliary structure used to represent the return value of `simpHaveTelescopeAux`.
+-/
+private structure SimpHaveResult where
+  /--
+  The simplified expression in `(fun x => b) v` form. Note that it may contain loose bound variables.
+  `simpHaveTelescope` attempts to minimize the quadratic overhead imposed
+  by the locally nameless discipline in a sequence of `have` expressions.
+  -/
+  expr     : Expr
+  /--
+  The type of `expr`. It may contain loose bound variables like in the `expr` field.
+  Used in proof construction.
+  -/
+  exprType : Expr
+  /--
+  The initial expression in `(fun x => b) v` form.
+  -/
+  exprInit   : Expr
+  /--
+  The expression `expr` in `have x := v; b` form.
+  -/
+  exprResult : Expr
+  /--
+  The proof that the simplified expression is equal to the input one.
+  It may contain loose bound variables like in the `expr` field.
+  -/
+  proof    : Expr
+  /--
+  `modified := false` iff `expr` is structurally equal to the input expression.
+  When false, `proof` is `Eq.refl`.
+  -/
+  modified : Bool
+
+private def SimpHaveResult.toResult (u : Level) (source : Expr) : SimpHaveResult → Result
+  | { expr, exprType, exprInit, exprResult, proof, modified, .. } =>
+    { expr := exprResult
+      proof? :=
+        if modified then
+          -- Add a type hint to convert back into `have` form.
+          some <| mkApp2 (mkConst ``id [levelZero]) (mkApp3 (mkConst ``Eq [u]) exprType source exprResult)
+            -- Add in a second type hint, for use in an optimization to avoid zeta/beta reductions in the kernel
+            -- (see the base case in `simpHaveTelescopeAux`).
```

Comment:
@kmill Could you please explain why two type hits are needed?
For example, what changes would be required in the kernel to eliminate this workaround?